### PR TITLE
Add various requested features

### LINKED ISSUE #179: add only keyword to set_global_backend
from the discussion in scipy/scipy#10383

>    should uarray have an "only" keyword for set_global_backend as well?

not sure whether or not this would be commonly needed, but it would prevent falling back to other registered backends if set.

### LINKED ISSUE #181: add a method to clear registered and/or global backends
from the discussion in scipy/scipy#10383

>   should there be a mechanism to clear registered backends and/or the global backend?

### PR DESCRIPTION
**Pull request by hameerabbasi**
This PR adds:

* Clearing of backends (Fixes #181)
* Only keyword to `set_global_backend` (Fixes #179)
* Tests for registration of backends.

### CODE REVIEW

File: uarray/_uarray_dispatch.cxx [hunk 1]
```diff
@@ -51,11 +51,6 @@ class py_ref
       std::swap(other.obj_, obj_);
     }
 
-  friend void swap(py_ref & a, py_ref & b) noexcept
```
Comment (peterbell10) [inline uarray/_uarray_dispatch.cxx]: why?

File: uarray/_uarray_dispatch.cxx [hunk 2]
```diff
@@ -455,7 +499,7 @@ LoopReturn for_each_backend(const std::string & domain_key, Callback call)
   LoopReturn ret = LoopReturn::Continue;
   for (int i = pref.size()-1; i >= 0; --i)
   {
-    auto options = pref[i];
+    auto& options = pref[i];
```
Comment (peterbell10) [inline uarray/_uarray_dispatch.cxx]: This will blow up if `call` modifies the list of preferred backends.

File: uarray/_uarray_dispatch.cxx [hunk 3]
```diff
@@ -338,7 +382,7 @@ struct SetBackendContext
       return 0;
     }
 
-  static PyObject * enter__(SetBackendContext * self, PyObject * /*args*/)
+  static PyObject * enter__(SetBackendContext * self)
```
Comment (peterbell10) [inline uarray/_uarray_dispatch.cxx]: This is wrong, see the [docs](https://docs.python.org/3/c-api/structures.html#c.PyCFunction)
> `PyCFunction`
    Type of the functions used to implement most Python callables in C. Functions of this type take two `PyObject*` parameters and return one such value.

File: uarray/_uarray_dispatch.cxx [hunk 4]
```diff
@@ -199,24 +193,34 @@ PyObject * clear_all_globals(PyObject * /*self*/, PyObject * /*args*/)
...
+  options.backend = py_ref::ref(backend);
+  options.coerce = coerce;
+  options.only = only;
+
+  global_domain_map[domain].global = options;
   Py_RETURN_NONE;
 }
 
-PyObject * register_backend(PyObject * /*self*/, PyObject * args)
+PyObject * register_backend(PyObject * Py_UNUSED(self), PyObject * args)
```
Comment (peterbell10) [inline uarray/_uarray_dispatch.cxx]: `Py_UNUSED` is _maybe_ unused, commenting out the name is compile error on use.

File: uarray/_uarray_dispatch.cxx [hunk 5]
```diff
@@ -227,6 +231,46 @@ PyObject * register_backend(PyObject * /*self*/, PyObject * args)
...
+    global_domain_map[domain].global.backend = py_ref::ref(nullptr);
+  }
+}
+
+PyObject * clear_backends(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  char* domain = nullptr;
+  int registered = true, global = false;
+  if (!PyArg_ParseTuple(args, "s|pp", 
+    &domain, &registered, &global))
```
Comment (peterbell10) [inline uarray/_uarray_dispatch.cxx]: This won't accept domains with null characters. Just take by Object and call `domain_to_string`.

Comment (hameerabbasi) [reply uarray/_uarray_dispatch.cxx]: Will domains have null characters?

File: uarray/_uarray_dispatch.cxx (hunk 4 above)
Comment (hameerabbasi) [reply uarray/_uarray_dispatch.cxx]: IIRC this also modifies the name so using it is a compiler error.

File: uarray/_uarray_dispatch.cxx (hunk 5 above)
Comment (peterbell10) [reply uarray/_uarray_dispatch.cxx]: It's an arbitrary string so there's no reason it couldn't.  Better to use `domain_to_string` everywhere either way.

File: uarray/_uarray_dispatch.cxx (hunk 3 above)
Comment (hameerabbasi) [reply uarray/_uarray_dispatch.cxx]: Interesting, I was following this: https://pythonextensionpatterns.readthedocs.io/en/latest/parsing_arguments.html#no-arguments

File: uarray/_uarray_dispatch.cxx (hunk 1 above)
Comment (hameerabbasi) [reply uarray/_uarray_dispatch.cxx]: I got a warning if it existed.

File: uarray/_uarray_dispatch.cxx (hunk 4 above)
Comment (peterbell10) [reply uarray/_uarray_dispatch.cxx]: https://github.com/python/cpython/blob/b49858b4b7b4c9d85ef6946ad020f83e4fa1caa7/Include/pymacro.h#L98

It adds `__attribute__((unused))` which as per GCC's docs
>This attribute, attached to a variable, means that the variable is meant to be **possibly unused**. GCC will not produce a warning for this variable. 

(Emphasis mine)

Comment (peterbell10) [reply uarray/_uarray_dispatch.cxx]: Actually, it does modify the name...
Still no reason to use it in C++.

File: uarray/_uarray_dispatch.cxx [hunk 6]
```diff
@@ -227,6 +231,46 @@ PyObject * register_backend(PyObject * /*self*/, PyObject * args)
   Py_RETURN_NONE;
 }
 
+void clear_single(std::string domain, bool registered, bool global)
```
Comment (peterbell10) [inline uarray/_uarray_dispatch.cxx]: `const std::string &` to avoid a copy.

File: uarray/_uarray_dispatch.cxx [hunk 7]
```diff
@@ -227,6 +231,46 @@ PyObject * register_backend(PyObject * /*self*/, PyObject * args)
...
+{
+  if (global_domain_map.find(domain) == global_domain_map.end())
```
Comment (peterbell10) [inline uarray/_uarray_dispatch.cxx]: Save the result from the first lookup and then you don't need to do repeated lookups.

File: uarray/_uarray_dispatch.cxx [hunk 8]
```diff
@@ -227,6 +231,46 @@ PyObject * register_backend(PyObject * /*self*/, PyObject * args)
...
+
+PyObject * clear_backends(PyObject * /* self */, PyObject * args)
+{
+  PyObject* domain = nullptr;
+  int registered = true, global = false;
+  if (!PyArg_ParseTuple(args, "O|pp", 
+    &domain, &registered, &global))
+    return nullptr;
+
+  if ((!domain || domain == Py_None)  && registered && global)
```
Comment (peterbell10) [inline uarray/_uarray_dispatch.cxx]: `domain` can't be `nullptr` here because it's a non-optional argument.

File: uarray/_uarray_dispatch.cxx [hunk 9]
```diff
@@ -190,7 +184,7 @@ std::string backend_to_domain_string(PyObject * backend)
  * This must be installed in a python atexit handler. This prevents Py_DECREF
  * being called after the interpreter has already shudown.
  */
-PyObject * clear_all_globals(PyObject * /*self*/, PyObject * /*args*/)
+PyObject * clear_all_globals(PyObject * /* self */)
```
Comment (peterbell10) [inline uarray/_uarray_dispatch.cxx]: This one's still wrong

File: uarray/_uarray_dispatch.cxx [hunk 10]
```diff
@@ -227,6 +231,48 @@ PyObject * register_backend(PyObject * /*self*/, PyObject * args)
...
+    }
+  
+  if (registered)
+  {
+    domain_globals->second.registered.clear();
+  }
+
+  if (global)
+  {
+    domain_globals->second.global.backend = py_ref::ref(nullptr);
```
Comment (peterbell10) [inline uarray/_uarray_dispatch.cxx]: Nit:
```suggestion
    domain_globals->second.global.backend.reset();
```

File: uarray/_uarray_dispatch.cxx [hunk 11]
```diff
@@ -338,7 +384,7 @@ struct SetBackendContext
       return 0;
     }
 
-  static PyObject * enter__(SetBackendContext * self, PyObject * /*args*/)
+  static PyObject * enter__(SetBackendContext * self, PyObject * /* self */)
```
Comment (peterbell10) [inline uarray/_uarray_dispatch.cxx]:
```suggestion
  static PyObject * enter__(SetBackendContext * self, PyObject * /* args */)
```

File: uarray/_backend.py [hunk 12]
```diff
@@ -250,6 +259,33 @@ def register_backend(backend):
...
+        Whether or not to clear registered backends. See :obj:`register_backend`.
+    globals : bool
+        Whether or not to clear global backends. See :obj:`set_global_backend`.
+    
+    See Also
+    --------
+    register_backend : Register a backend globally.
+    set_global_backend : Set a global backend.
+    """
+    _uarray.clear_backends(domain, registered, globals)
```
Comment (peterbell10) [inline uarray/_backend.py]: It's probably worth mentioning that the program is likely to crash if we're already inside a uarray method call.

**Comment by grlee77** [comment]
Thanks @hameerabbasi, that was pretty quick!

**Pull request merged**